### PR TITLE
Allow attributes on expression

### PR DESCRIPTION
**Pull request by pcpthm**
This PR adds a new AST node `AttrExpr`. This node is used when outer attributes are applied to an expression that is not an expression statement.
```rust
fn foo() {
    bar(#[attr] 1); // CallExpr(.. ArgList(.. AttrExpr(..)))
    #[attr] func();   // ExprStmt(Attr(..), CallExpr(..))
    #[attr] func()    // AttrExpr(Attr(..), CallExpr(..)) last expression of a block is not an ExprStmt
}
```

It seems like stable rustc allows attributes on certain expressions when operator precedence is not confusing despite the unstable [stmt_expr_attributes] feature.
This PR currently isn't trying to produce parse errors and instead allowing attributes on arbitrary expression with the same precedence as lambda (this is suggested in the [stmt_expr_attributes] tracking issue). Probably errors can be reported on a validation pass.

This PR addresses #919's second and third case.

[stmt_expr_attributes]: https://github.com/rust-lang/rust/issues/15701

### CODE REVIEW

File: crates/ra_parser/src/grammar/expressions.rs [hunk 1]
```diff
@@ -453,8 +445,13 @@ pub(crate) fn named_field_list(p: &mut Parser) {
...
     while !p.at(EOF) && !p.at(R_CURLY) {
         match p.current() {
-            IDENT => {
+            // test struct_literal_field_with_attr
+            // fn main() {
+            //     S { #[cfg(test)] field: 1 }
+            // }
+            IDENT | POUND => {
                 let m = p.start();
+                attributes::outer_attributes(p);
```
Comment (matklad) [inline crates/ra_parser/src/grammar/expressions.rs]: Let's maybe send this change as a separate PR, which can be r+ed immediately?

**Comment by matklad** [comment]
Hm, this `expr_block_contents` looks way to complicated. Let me see if I can simplify it a bit...

**Comment by pcpthm** [comment]
I did that refactoring. Will update PR.

**Comment by matklad** [comment]
@pcpthm would you mind rebasing this on a copule of cosmetic changes I am about to push?

**Comment by matklad** [comment]
(basically, flattening this all a bit)

**Comment by pcpthm** [comment]
@matklad I rebased your commit. This PR is updated to include bigger change than first (a new AST syntax kind) to address issue on attributes on non-`ExprStmt` expression leaking to outer block. Do you think introducing a new AST node right approach?

**Comment by matklad** [comment]
>Do you think introducing a new AST node right approach?

I don't think so, treating attrs as children of expressions seems better. Ie, we don't create "decorator" nodes, for things like modifiers or visibilty, we store them directly as children and introduce traits like "visibility owner".

**Comment by matklad** [comment]
Whet we need here I think is an `Option<Marker>` arguemnt  to `expr_stmt`. See, for example, how `block_expr` works

**Comment by pcpthm** [comment]
Actually, I couldn't come up with how to implement parsing without introducing an another node because each expression type start an own node and I can not inject children to such node. If there is a way to "meld" two nodes after parsing, it can be much simpler to implement, I think.

**Comment by pcpthm** [comment]
@matklad The problem is `Option<Marker>` needs to be passed to not `expr_stmt`, every expression types possible (`tuple_expr`, `array_expr`, ... etc. ) if I am correct. That is a modularity hazard and want to avoid.

**Comment by matklad** [comment]
>. If there is a way to "meld" two nodes after parsing, it can be much simpler to implement, I think.

We can do this (`CompletedMarker::precede` achieves something similar). I am not sure it would be the definitive win, we need to try things out and make sure we use the single approach everywhere.

Currently, at least two approaches are used simulateneosly:

* don't call start/finish at all inside parsing function, let the caller take care of this (`maybe_item` works like this)
* instead of calling `start`, accept a `Marker` argument call finish explicitelly (`lifetime_param` does this)

And there's a third one:
* call both start/finish, but allow caller to prepend values to the nodes (this'll require new API)

My gut feeling is that the `Marker` approach, although a bit boilerplatey, is the most robust one.

**Comment by matklad** [comment]
Another thing to keep in mind is that if we want to support parsing attributes "inside" expressions, like

```
1 + #[foo] 2
```

We'll need t teach `atom` to parse attrs anyway

**Comment by killercup** [comment]
Am I right to assume this also addresses #759?

**Comment by pcpthm** [comment]
@killercup
Attributes on match arms are already handled (no error) but parse tree seems incorrect.

**Comment by killercup** [comment]
Ah, cool! Updated the issue to show the second case, attributes on blocks, don't work yet :)

**Comment by matklad** [comment]
So, how about this plan:

* first, we refactor `maybe_item` to also accept `Marker` as an argument: the current approach is confusing, and doing this refactor should give us a rough feeling about how it'd work for expressions
* if it works ok for items, we could try it for expressions as well

I wouldn't worry too much about modularity here parser is nicely isolated from the rest of the analyzer, and this is what is most important.

**Comment by pcpthm** [comment]
Okay, let me try that plan. My word about modularity was too much and actually I'm thinking that approach is not bad after thinking.

**Comment by pcpthm** [comment]
Another PR with suggested approach incoming.

**Comment by Boscop** [comment]
I just installed and enabled it, but it gives me "false" errors like this: 

image

In main.rs, I have `stmt_expr_attributes` and `type_ascription` enabled (among many other features). How can I tell rust-analyzer to take my enabled features into account? :)
